src/test/lib/EventTest.cpp

```diff
@@ -395,7 +395,7 @@ QuicTestValidateConnectionEvents1(
     ConnValidator Server(
         new(std::nothrow) ConnEventValidator* [6] {
             new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_CONNECTED, 0, true),
-            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_SHUTDOWN_INITIATED_BY_PEER),
+            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_SHUTDOWN_INITIATED_BY_TRANSPORT),
```

Comment:
These and other changes are most likely inevitable. The affected scenario is clients closing the connections from inside `QUIC_CONNECTION_EVENT_CONNECTED` callback. (doing so from server still sends app close).

The reason is that RFC suggests that before handshake is **confirmed** (which for client means receiving HANDSHAKE_COMPLETE frame from server), endpoints send CONNECTION CLOSE also on lower protection levels. However, the `QUIC_CONNECTION_EVENT_CONNECTED` is signaled immediately after client-side handshake *completes* (meaning TLS layer finishes processing remote Finished and produces local Finished TLS messages).

To keep the old behavior, we would likely have to either
- delay the event until handshake is confirmed which would be more disrupting change as it would add latency to connection establishment, or
- somehow send the app close only, and only after some time send the CONNECTION_CLOSE on lower levels in case peer is "stuck" and does not process the 1-RTT frame with app close. This sounds rather difficult given the code architecture and feels rather brittle.

Reply:
Let's just take (and backport) this change. For the backport, we will likely need to update this to allow for either event (making them both optional).